graphql_api/validation.py

```diff
@@ -0,0 +1,65 @@
+from typing import Any, Type
+
+from graphql import GraphQLError, ValidationRule
+from graphql.language.ast import DocumentNode, FieldNode, OperationDefinitionNode
+from graphql.validation import ValidationContext
+
+
+def create_max_depth_rule(max_depth: int) -> Type[ValidationRule]:
+    class MaxDepthRule(ValidationRule):
+        def __init__(self, context: ValidationContext) -> None:
+            super().__init__(context)
+            self.operation_depth: int = 1
+            self.max_depth_reached: bool = False
+            self.max_depth: int = max_depth
+
+        def enter_operation_definition(
```

Comment:
just curious, are these functions base functions you had to override default behavior of?

Similar story with enter_field, leave_field, and enter_document

Reply:
Yup! Though I don't think about it so much as "override" as "implement the interface" that is defined [here](https://github.com/graphql-python/graphql-core/blob/46244446c66ca3033f5295f6c796108e4c1f7365/src/graphql/language/visitor.py#L71). The function names `enter_X` and `leave_X` are dynamic per [here](https://github.com/graphql-python/graphql-core/blob/46244446c66ca3033f5295f6c796108e4c1f7365/src/graphql/language/visitor.py#L147C12-L150C60). And for any method that's not implemented explicitly, it behaves as a no-op ([here](https://github.com/graphql-python/graphql-core/blob/46244446c66ca3033f5295f6c796108e4c1f7365/src/graphql/language/visitor.py#L280C13-L281C30)).

this is the stuff I looked at in Ariadne [doc](https://ariadnegraphql.org/docs/query-validators#implementing-custom-validator) & this [example implementation](https://github.com/mirumee/ariadne/blob/af244283987ef57c186a512a1186f819f4dd6e20/ariadne/validation/query_cost.py#L44)

Reply:
Very cool! Thanks for linking all those docs, that was a fun set of reads